### PR TITLE
Add flush job mode to ingester.

### PR DESCRIPTION
**Pull request by gouthamve**
This is my attempt to simplify things compared to #1747

/cc @jtlisi @codesome

Signed-off-by: Goutham Veeramachaneni <gouthamve@gmail.com>

**What this PR does**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/cortex/cortex.go [hunk 1]
```diff
@@ -238,7 +241,20 @@ func (t *Cortex) initModule(cfg *Config, m moduleName) error {
 
 // Run starts Cortex running, and blocks until a signal is received.
 func (t *Cortex) Run() error {
-	return t.server.Run()
+	var err error
+	done := make(chan struct{})
```
Comment (codesome) [inline pkg/cortex/cortex.go]: We are not waiting for this anywhere (I will create a follow-up PR to your branch, just a note to myself)

**Comment by gouthamve** [comment]
> This might create it's own problem

Can you think of cases? I see this flush-job as something you run after the cluster has stabilised and I don't see any issues with it. Even if the cluster is in flux, I don't see it causing an major issues.

> There is a long wait before we start flushing even when we know it's a flush job

How / Where?

**Comment by codesome** [comment]
> you run after the cluster has stabilised

If we assume this then should be fine. But it is still an assumption. I am afraid it might interfere with transfer otherwise? I prefer running this job without worrying about the state of the cluster, as that just adds another dependency for the job.

> How / Where?

Apparently it does not after the `TransferOut` is disabled :)

File: pkg/cortex/cortex.go [hunk 2]
```diff
@@ -238,7 +241,22 @@ func (t *Cortex) initModule(cfg *Config, m moduleName) error {
...
+	go func() {
+		err = t.server.Run()
+
+		close(done)
+	}()
+
+	// Initiate shutdown if its a job to flush data.
+	if t.cfg.Ingester.IsFlushJob {
+		_ = t.Stop()
```
Comment (codesome) [inline pkg/cortex/cortex.go]: We need to make sure that calling `t.Stop()` twice is not a problem. Currently, `main.go` would also call this and try stopping the modules again.

**Comment by codesome** [comment]
I have rebased https://github.com/cortexproject/cortex/pull/1747 on top of this with a small fix. I think we can close this now.

**Comment by gouthamve** [comment]
Closing this as most of the changes have been incorporated in #1747

**Pull request closed without merging**